packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useExternalPlugins.ts

```diff
@@ -0,0 +1,91 @@
+import { useEffect, useState } from 'react';
+import { PluginStates } from '@joplin/lib/services/plugins/reducer';
+import { contentScriptsToCodeMirrorPlugin } from '@joplin/lib/services/plugins/utils/loadContentScripts';
+import { extname } from 'path';
+import shim from '@joplin/lib/shim';
+import uuid from '@joplin/lib/uuid';
+
+const { reg } = require('@joplin/lib/registry.js');
+
+export default function useExternalPlugins(CodeMirror: any, plugins: PluginStates) {
+
+	const [options, setOptions] = useState({});
+	useEffect(() => {
+		let newOptions = {};
+
+		const contentScripts = contentScriptsToCodeMirrorPlugin(plugins);
+
+		for (const contentScript of contentScripts) {
+			try {
+				const mod = contentScript.module;
+
+				if (mod.codeMirrorAssets) {
+					for (const asset of mod.codeMirrorAssets) {
+						try {
+							require(`codemirror/${asset}`);
+						} catch (error) {
+							error.message = `${asset} is not a valid CodeMirror asset, keymap or mode. You can find a list of valid assets here: https://codemirror.net/doc/manual.html#addons`;
+							throw error;
+						}
+					}
+				}
+
+				if (mod.codeMirrorOptions) {
+					newOptions = Object.assign({}, newOptions, mod.codeMirrorOptions);
+				}
+
+				if (mod.assets) {
+					const cssStrings = [];
+
+					for (const asset of mod.assets()) {
+						let mime = asset.mime;
+
+						if (!mime && asset.inline) throw new Error('Mime type is required for inline assets');
+
+						if (!mime && asset.name) {
+							const ext = extname(asset.name).toLowerCase();
+							if (ext === '.css') mime = 'text/css';
+						}
+
+						if (mime !== 'text/css') throw new Error('Only css assets are supported for CodeMirror plugins');
+
+						if (asset.inline) {
+							cssStrings.push(asset.text);
+						} else {
+							addScript(shim.fsDriver().resolveRelativePathWithinDir(contentScript.assetPath, asset.name), contentScript.id);
+						}
+					}
+
+					if (cssStrings.length > 0) {
+						addInlineCss(cssStrings, contentScript.id);
+					}
+				}
+
+				if (mod.plugin) {
+					mod.plugin(CodeMirror);
+				}
+			} catch (error) {
+				reg.logger().error(error.toString());
+				continue;
```

Comment:
Just a small thing but I think this `continue` could be removed?

Reply:
I put that there in case the next person editing the file placed code outside the try/catch block, in that case the `continue` would be necessary to ensure correct behaviour. In hindsight we shouldn't place anything outside the try/catch block so this won't be necessary (and won't be helpful).